web/old/remctl-2.14/docs/remctld.8
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "REMCTLD 8"
.TH REMCTLD 8 "2009-05-22" "2.14" "remctl"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
remctld \- Server for remctl, a remote command execution utility
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
remctld [\fB\-dFhmSv\fR] [\fB\-f\fR \fIconfig\fR] [\fB\-k\fR \fIkeytab\fR] [\fB\-P\fR \fIfile\fR]
[\fB\-p\fR \fIport\fR] [\fB\-s\fR \fIservice\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBremctld\fR is the server for remctl.  It accepts a connection from remctl,
receives the command to execute and the arguments, verifies authorization
of the user and executes the command, returning the result back to the
client.  All connections are authenticated using GSS-API Kerberos v5, and
all transmissions are also encrypted using the GSS-API privacy layer.
.PP
\&\fBremctld\fR is normally started using \fBtcpserver\fR or from \fBinetd\fR, but it
may be run in stand-alone mode as a daemon using \fB\-m\fR.  Either \fB\-s\fR must
be given to use an alternate identity (which will require the same flag be
used for \fBremctl\fR client invocations), or it must be run as root to read
the host keytab file.  \fBremctld\fR logs its activity using syslog (the
daemon facility).
.PP
The location of the configuration file may be specified with the \fB\-f\fR
option.  The default location is \fI/usr/local/etc/remctl.conf\fR.  For
information on the format of the configuration file, see \*(L"\s-1CONFIGURATION\s0
\&\s-1FILE\s0\*(R" below.
.PP
When the command is run, several environment variables will be set
providing information about the remote connection.  See \s-1ENVIRONMENT\s0
below for more information.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-d\fR" 4
.IX Item "-d"
Enable verbose debug logging to syslog (or to standard output if \fB\-S\fR is
also given).
.IP "\fB\-F\fR" 4
.IX Item "-F"
Normally when running in stand-alone mode (\fB\-m\fR), \fBremctld\fR backgrounds
itself to run as a daemon, changes directory to \fI/\fR, and drops any
controlling terminal.  This flag suppresses this behavior, usually for
debugging or so that \fBremctld\fR can be monitored by other processes.
.IP "\fB\-f\fR \fIconfig\fR" 4
.IX Item "-f config"
The configuration file for \fBremctld\fR, overriding the default path.
.IP "\fB\-h\fR" 4
.IX Item "-h"
Show a brief usage message and then exit.
.IP "\fB\-k\fR \fIkeytab\fR" 4
.IX Item "-k keytab"
Use \fIkeytab\fR as the keytab for server credentials rather than the system
default or the value of the \s-1KRB5_KTNAME\s0 environment variable.  Using \fB\-k\fR
just sets the \s-1KRB5_KTNAME\s0 environment variable internally in the process.
.IP "\fB\-m\fR" 4
.IX Item "-m"
Enable stand-alone mode.  \fBremctld\fR will listen to its configured port
and fork a new child for each incoming connection.  By default, when this
option is used, \fBremctld\fR also changes directory to \fI/\fR, backgrounds
itself, and closes standard input, output, and error.  To not background,
pass \fB\-F\fR as well.  To not close standard output and error and continue
using them for logging, pass \fB\-S\fR as well.
.Sp
To determine the port, \fBremctld\fR attempts to look up the \f(CW\*(C`remctl\*(C'\fR
service in the local \fI/etc/services\fR file and uses the port defined
there.  If the \f(CW\*(C`remctl\*(C'\fR service could not be found, it uses 4373, the
registered remctl port.
.IP "\fB\-P\fR \fIfile\fR" 4
.IX Item "-P file"
When running in stand-alone mode (\fB\-m\fR), write the \s-1PID\s0 of \fBremctld\fR to
\&\fIfile\fR.  This option is ignored unless \fB\-m\fR is also given.
.IP "\fB\-p\fR \fIport\fR" 4
.IX Item "-p port"
When running in stand-alone mode, Listen on port \fIport\fR rather than the
default.  This option does nothing unless used with \fB\-m\fR.
.IP "\fB\-S\fR" 4
.IX Item "-S"
Rather than logging to syslog, log debug and routine connection messages
to standard output and error messages to standard error.  This option is
mostly useful for testing and debugging.
.IP "\fB\-s\fR \fIservice\fR" 4
.IX Item "-s service"
Specifies which principal is used as the server identity for client
authentication.  The client must also use the same identity as the server
identity for authentication to succeed.  By default, \fBremctld\fR accepts
any principal with a key in the default keytab file (which can be changed
with the \fB\-k\fR option).  This is normally the most desirable behavior.
.IP "\fB\-v\fR" 4
.IX Item "-v"
Print the version of \fBremctld\fR and exit.
.SH "CONFIGURATION FILE"
.IX Header "CONFIGURATION FILE"
The configuration file defines the allowed commands and specifies access
control information.  The configuration file format is lines of space\- or
tab-separated strings, where each line is:
.PP
.Vb 1
\&    command subcommand executable [option=value ...] acl [acl ...]
.Ve
.PP
Each command consists of a command, a subcommand, and zero or more
arguments.  Each configuration line defines an acceptable command and
subcommand (or, if \f(CW\*(C`ALL\*(C'\fR is used as mentioned below under \fIservice\fR, a
set of commands).  The first configuration line matching the received
command is used, so list more specific entries before more general
entries.
.PP
Blank lines and lines beginning with \f(CW\*(C`#\*(C'\fR are ignored.  Lines can be
continued on the next line by ending them with a backslash (\f(CW\*(C`\e\*(C'\fR).  Be
aware that comments can be continued with a backslash as well.
.PP
As a special case, a line like:
.PP
.Vb 1
\&    include file
.Ve
.PP
will include \fIfile\fR as if its contents were pasted verbatim into the
configuration file at that point.  \fIfile\fR may be a directory, in which
case all files whose names do not contain a period found in that directory
will be included (in no particular order).  \fIfile\fR should be a fully
qualified path.
.PP
The meaning of these fields is:
.IP "\fIcommand\fR" 4
.IX Item "command"
The command being issued.  Normally, related commands (such as all
commands for managing a particular service) are grouped together as
subcommands under one command.
.IP "\fIsubcommand\fR" 4
.IX Item "subcommand"
The subcommand within the command being requested, such as \f(CW\*(C`release\*(C'\fR for
the release function of the \s-1AFS\s0 volume backend.  If the keyword \f(CW\*(C`ALL\*(C'\fR is
used instead of a specific subcommand, this line matches all subcommands
with the given command and can be used to dispatch all subcommands under
that command to the same executable with the same ACLs.  Since the first
matching entry is used, list entries for specific services first (if any)
and then the \f(CW\*(C`ALL\*(C'\fR catch-all.
.Sp
\&\fBremctld\fR accepts commands with no subcommand argument.  Such commands
will only match lines in the configuration file with \f(CW\*(C`ALL\*(C'\fR for the
service.  (The \fBremctl\fR client will not send such commands, but they can
be sent using the client library or via another implementation.)
.Sp
The subcommand is always passed as the first argument to the executable
program that is listed for that service unless no subcommand was given.
.IP "\fIexecutable\fR" 4
.IX Item "executable"
The full path to the command executable to run for this command and
subcommand combination.  (See examples below.)
.IP "\fIoption\fR=\fIvalue\fR" 4
.IX Item "option=value"
An option setting that applies to this command.  Supported option settings
are:
.RS 4
.IP "\fIlogmask\fR=\fIn\fR[,...]" 4
.IX Item "logmask=n[,...]"
Limit logging of command arguments.  Any argument listed in the logmask
list will have its value logged as \*(L"**MASKED**\*(R".  This is to avoid logging
the arguments of commands that take private information such as passwords.
The logmask list should contain argument numbers separated by commas, with
the \fIsubcommand\fR considered argument 1.  The \fIcommand\fR argument cannot
be masked.
.Sp
For example, if the command is \f(CW\*(C`admin passwd \f(CIusername\f(CW \f(CIpassword\f(CW\*(C'\fR,
then you'd want to set logmask to \f(CW3\fR, so the password argument gets
logged as \f(CW\*(C`**MASKED**\*(C'\fR.  If the command is \f(CW\*(C`user passwd \f(CIusername\f(CW
\&\f(CIold\-password\f(CW \f(CInew\-password\f(CW\*(C'\fR, you'd want to set logmask to \f(CW\*(C`3,4\*(C'\fR.
.ie n .IP "\fIstdin\fR=(\fIn\fR | ""last"")" 4
.el .IP "\fIstdin\fR=(\fIn\fR | \f(CWlast\fR)" 4
.IX Item "stdin=(n | last)"
Specifies that the \fIn\fRth or last argument to the command be passed on
standard input instead of on the command line.  The value of this option
must either be the number of argument to pass on standard input (with the
\&\fIsubcommand\fR considered argument 1) or the special value \f(CW\*(C`last\*(C'\fR, which
indicates that the final argument (no matter how many there are) be passed
on standard input.
.Sp
The \fIsubcommand\fR cannot be passed on standard input, so \fIn\fR must be at
least \f(CW2\fR.  If this option is set to \f(CW\*(C`last\*(C'\fR and no arguments are given
except the \fIcommand\fR and possibly the \fIsubcommand\fR, nothing will be
passed on standard input.
.Sp
This option is used primarily for passing large amounts of data that may
not fit on the command line or data that contains \s-1NUL\s0 characters.  It can
also be used for arguments like passwords that shouldn't be exposed on the
command line.  Only at most one argument may be passed on standard input
to the command.
.RE
.RS 4
.RE
.IP "\fIacl\fR" 4
.IX Item "acl"

debian/changelog
gutenbach (2.10) unstable; urgency=low

  * Add a useful reminder message to client scripts, "Did you forget to add
    it [the queue] with gutenbach-client-config?"
  * Add 'last' keyword for gbr
  * Use perl bindings for gbr and gbrm in addition to gbq.
  * Update README
  * Add support for pathname expansion/multiple files in gbr
  * Added the "current" option to gbrm, which just removes the current job
    from the queue

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 06 Jul 2010 00:11:23 -0400

gutenbach (2.09) unstable; urgency=low

  * Fixed support for playing YouTube videos
  * Fixed error with CUPS_SERVER in gbr and gbrm
  * Added supportfor playing YouTube URLs with gbr

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 05 Jul 2010 15:32:18 -0400

gutenbach (2.08) unstable; urgency=low

  * Remove deprecated lpd files in gutenbach-server
  * Modify gutenbach-queue to use /tmp instead of /var/run
  * Replace cupsd.conf with our own version, backing it up to
    cupsd.conf.gutenbach-orig in gutenbach-server
  * Modify gbr and gbrm to use CUPS_SERVER instead of -h in
    gutenbach-client

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 29 Jun 2010 19:35:58 -0400

gutenbach (2.07) unstable; urgency=low

  * Update gutenbach-queue to be compatible with CUPS

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 29 Jun 2010 01:38:34 -0400

gutenbach (2.06) unstable; urgency=low

  * License gutenbach under GPLv2

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 28 Jun 2010 22:46:42 -0400

gutenbach (2.05) unstable; urgency=low

  * Add manpages for gutenbach-client scripts
  * Fix incorrect path in gutenbach-server.install
  * Rename gutenbach to gutenbach-server in po

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 28 Jun 2010 22:23:15 -0400

gutenbach (2.04) unstable; urgency=low

  * Create a single source package for all gutenbach-* packages.

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 20:53:20 -0400

gutenbach-server (2.03) unstable; urgency=low

  * Update permissions again, to allow everybody to access the server

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 08 Jun 2010 19:07:09 -0400

gutenbach-server (2.02) unstable; urgency=low

  * Fix print server permissions problems such that remote
  users can user cups-lpq and cups-lprm

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 31 May 2010 23:51:52 -0400

gutenbach-server (2.01) unstable; urgency=low

  * Handle SIGTERM in the gutenbach backend, so jobs don't
  keep playing even after being dequeued.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 31 May 2010 22:23:53 -0400

gutenbach-server (2.0) unstable; urgency=low

  * Changed from using LPRng to CUPS

 -- Jessica Hamrick <jhamrick@mit.edu>  Thu, 13 May 2010 17:40:53 -0400

gutenbach-server (1.01) unstable; urgency=low

  * Implemented debconf to retrieve printer and zephyr class names.
  * Made uninstall cleaner (and when you purge, it purges
  configuration files and changes)

 -- Jessica Hamrick <jhamrick@mit.edu>  Wed, 16 Sep 2009 15:58:40 -0400

gutenbach-server (1.0) unstable; urgency=low

  * Initial release.

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 08 Sep 2009 18:38:12 -0400

gutenbach-client (1.01-1) unstable; urgency=low

  * Make package no longer Debian native.

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 19:20:08 -0400

gutenbach-client (1.01) unstable; urgency=low

  * Fix debhelper compatibility to 5 (for backwards compatibility)

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 18:01:02 -0400

gutenbach-client (1.0) unstable; urgency=low

  * Initial package build
  * gbq - show queue
  * gbr - queue file
  * gbrm - remove file
  * gutenbach-client-config - add, list, or delete queues

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 16:56:28 -0400

gutenbach-client (0.01) unstable; urgency=low

  * Created skeleton package.

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 19 Jun 2010 00:23:53 -0400

gutenbach-queue (1.02) unstable; urgency=low

  * Prevent the screen from turning off when
  view-gutenbach-queue is running.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 20:24:37 -0500

gutenbach-queue (1.01) unstable; urgency=low

  * Changed the chown in the postinst to
  daemon:daemon instaed of daemon:pika

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 20:23:50 -0500

gutenbach-queue (1.0) unstable; urgency=low

  * Initial release.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 20:03:48 -0500

gutenbach-remctl (1.02) unstable; urgency=low

  * Depend on version 2.0 or greater of gutenbach

 -- Jessica Hamrick <jhamrick@mit.edu>  Thu, 13 May 2010 17:42:53 -0400

gutenbach-remctl (1.01) unstable; urgency=low

  * Just say the new volume, not who changed it.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 22:04:43 -0500

gutenbach-remctl (1.0) unstable; urgency=low

  * Initial release.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 22:03:44 -0500


web/old/remctl-2.14/server/remctld.c
/*
 * The remctl server.
 *
 * Handles option parsing, network setup, and the basic processing loop of the
 * remctld server.  Supports either being run from inetd or tcpserver or
 * running as a stand-alone daemon and managing its own network connections.
 *
 * Written by Anton Ushakov
 * Extensive modifications by Russ Allbery <rra@stanford.edu>
 * Copyright 2002, 2003, 2004, 2005, 2006, 2007, 2008
 *     Board of Trustees, Leland Stanford Jr. University
 *
 * See LICENSE for licensing terms.
 */

#include <config.h>
#include <portable/system.h>
#include <portable/gssapi.h>
#include <portable/socket.h>

#include <signal.h>
#include <syslog.h>
#include <sys/wait.h>
#include <time.h>

#include <server/internal.h>
#include <util/util.h>

/*
 * Flag indicating whether we've received a SIGCHLD and need to reap children
 * (only used in standalone mode).
 */
static volatile sig_atomic_t child_signaled = 0;

/*
 * Flag indicating whether we've received a signal asking us to re-read our
 * configuration file (only used in standalone mode).
 */
static volatile sig_atomic_t config_signaled = 0;

/*
 * Flag indicating whether we've received a signal asking us to exit (only
 * used in standalone mode).
 */
static volatile sig_atomic_t exit_signaled = 0;

/* Usage message. */
static const char usage_message[] = "\
Usage: remctld <options>\n\
\n\
Options:\n\
    -d            Log verbose debugging information\n\
    -f <file>     Config file (default: " CONFIG_FILE ")\n\
    -h            Display this help\n\
    -m            Stand-alone daemon mode, meant mostly for testing\n\
    -P <file>     Write PID to file, only useful with -m\n\
    -p <port>     Port to use, only for standalone mode (default: 4373)\n\
    -S            Log to standard output/error rather than syslog\n\
    -s <service>  Service principal to use (default: host/<host>)\n\
    -v            Display the version of remctld\n";

/* Structure used to store program options. */
struct options {
    bool foreground;
    bool standalone;
    bool log_stdout;
    bool debug;
    unsigned short port;
    char *service;
    const char *config_path;
    const char *pid_path;
};


/*
 * Display the usage message for remctld.
 */
static void
usage(int status)
{
    fprintf((status == 0) ? stdout : stderr, usage_message);
    if (status == 0)
        exit(0);
    else
        die("invalid usage");
}


/*
 * Signal handler for child processes forked when running in standalone mode.
 * Just set the child_signaled global so that we know to reap the processes
 * later.
 */
static RETSIGTYPE
child_handler(int sig UNUSED)
{
    child_signaled = 1;
}


/*
 * Signal handler for signals asking us to re-read our configuration file when
 * running in standalone mode.  Set the config_signaled global so that we do
 * this the next time through the processing loop.
 */
static RETSIGTYPE
config_handler(int sig UNUSED)
{
    config_signaled = 1;
}


/*
 * Signal handler for signals asking for a clean shutdown when running in
 * standalone mode.  Set the exit_signaled global so that we exit cleanly the
 * next time through the processing loop.
 */
static RETSIGTYPE
exit_handler(int sig UNUSED)
{
    exit_signaled = 1;
}


/*
 * Given a service name, imports it and acquires credentials for it, storing
 * them in the second argument.  Returns true on success and false on failure,
 * logging an error message.
 *
 * Normally, you don't want to do this; instead, normally you want to allow
 * the underlying GSS-API library choose the appropriate credentials from a
 * keytab for each incoming connection.
 */
static bool
acquire_creds(char *service, gss_cred_id_t *creds)
{
    gss_buffer_desc buffer;
    gss_name_t name;
    OM_uint32 major, minor;

    buffer.value = service;
    buffer.length = strlen(buffer.value) + 1;
    major = gss_import_name(&minor, &buffer, GSS_C_NT_USER_NAME, &name);
    if (major != GSS_S_COMPLETE) {
        warn_gssapi("while importing name", major, minor);
        return false;
    }
    major = gss_acquire_cred(&minor, name, 0, GSS_C_NULL_OID_SET,
                             GSS_C_ACCEPT, creds, NULL, NULL);
    if (major != GSS_S_COMPLETE) {
        warn_gssapi("while acquiring credentials", major, minor);
        return false;
    }
    gss_release_name(&minor, &name);
    return true;
}


/*
 * Handle the interaction with the client.  Takes the client file descriptor,
 * the server configuration, and the server credentials.  Establishes a
 * security context, processes requests from the client, checks the ACL file
 * as appropriate, and then spawns commands, sending the output back to the
 * client.  This function only returns when the client connection has
 * completed, either successfully or unsuccessfully.
 */
static void
server_handle_connection(int fd, struct config *config, gss_cred_id_t creds)
{
    struct client *client;

    /* Establish a context with the client. */
    client = server_new_client(fd, creds);
    if (client == NULL) {
        close(fd);
        return;
    }
    debug("accepted connection from %s (protocol %d)", client->user,
          client->protocol);

    /*
     * Now, we process incoming commands.  This is handled differently
     * depending on the protocol version.  These functions won't exit until
     * the client is done sending commands and we're done replying.
     */
    if (client->protocol == 1)
        server_v1_handle_commands(client, config);
    else
        server_v2_handle_commands(client, config);

    /* We're done; shut down the client connection. */
    server_free_client(client);
}


/*
 * Gather information about an exited child and log an appropriate message.
 * We keep the log level to debug unless something interesting happened, like
 * a non-zero exit status.
 */
static void
server_log_child(pid_t pid, int status)
{
    if (WIFEXITED(status)) {
        if (WEXITSTATUS(status) != 0)
            warn("child %lu exited with %d", (unsigned long) pid,
                 WEXITSTATUS(status));
        else
            debug("child %lu done", (unsigned long) pid);
    } else if (WIFSIGNALED(status)) {
        warn("child %lu died on signal %d", (unsigned long) pid,
             WTERMSIG(status));
    } else {
        warn("child %lu died", (unsigned long) pid);
    }
}


/*
 * Run as a daemon.  This is the main dispatch loop, which listens for network
 * connections, forks a child to process each connection, and reaps the
 * children when they're done.  This is only used in standalone mode; when run
 * from inetd or tcpserver, remctld processes one connection and then exits.
 */
static void
server_daemon(struct options *options, struct config *config,
              gss_cred_id_t creds)
{
    int s, stmp, status;
    pid_t child;
    struct sigaction sa, oldsa;
    struct sockaddr_storage ss;
    socklen_t sslen;
    char ip[INET6_ADDRSTRLEN];

    /* We're running as a daemon, so don't self-destruct. */
    alarm(0);

    /* Set up a SIGCHLD handler so that we know when to reap children. */
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = child_handler;
    if (sigaction(SIGCHLD, &sa, &oldsa) < 0)
        sysdie("cannot set SIGCHLD handler");

    /* Set up exit handlers for signals that call for a clean shutdown. */
    sa.sa_handler = exit_handler;
    if (sigaction(SIGINT, &sa, NULL) < 0)
        sysdie("cannot set SIGINT handler");
    if (sigaction(SIGTERM, &sa, NULL) < 0)
        sysdie("cannot set SIGTERM handler");

    /* Set up a SIGHUP handler so that we know when to re-read our config. */
    sa.sa_handler = config_handler;
    if (sigaction(SIGHUP, &sa, NULL) < 0)
        sysdie("cannot set SIGHUP handler");

    /* Log a starting message. */
    notice("starting");

    /* Bind to the network socket. */
    stmp = network_bind_ipv4("any", options->port);
    if (stmp < 0)
        sysdie("cannot create socket");
    if (listen(stmp, 5) < 0)
        sysdie("error listening on socket");

    /*
     * The main processing loop.  Each time through the loop, check to see if
     * we need to reap children, check to see if we should re-read our
     * configuration, and check to see if we're exiting.  Then see if we have
     * a new connection, and if so, fork a child to handle it.
     *
     * Note that there are no limits here on the number of simultaneous
     * processes, so you may want to set system resource limits to prevent an
     * attacker from consuming all available processes.
     */
    do {
        if (child_signaled) {
            child_signaled = 0;
            while ((child = waitpid(0, &status, WNOHANG)) > 0)
                server_log_child(child, status);
            if (child < 0 && errno != ECHILD)
                sysdie("waitpid failed");
        }
        if (config_signaled) {
            config_signaled = 0;
            notice("re-reading configuration");
            server_config_free(config);
            config = server_config_load(options->config_path);
            if (config == NULL)
                die("cannot load configuration file %s", options->config_path);
        }
        if (exit_signaled) {
            notice("signal received, exiting");
            if (options->pid_path != NULL)
                unlink(options->pid_path);
            exit(0);
        }
        sslen = sizeof(ss);
        s = accept(stmp, (struct sockaddr *) &ss, &sslen);
        if (s < 0) {
            if (errno != EINTR)
                syswarn("error accepting connection");
            continue;
        }
        fdflag_close_exec(s, true);
        child = fork();
        if (child < 0) {
            syswarn("forking a new child failed");
            warn("sleeping ten seconds in the hope we recover...");
            sleep(10);
        } else if (child == 0) {
            close(stmp);
            if (sigaction(SIGCHLD, &oldsa, NULL) < 0)
                syswarn("cannot reset SIGCHLD handler");
            server_handle_connection(s, config, creds);
            if (options->log_stdout)
                fflush(stdout);
            exit(0);
        } else {
            close(s);
            network_sockaddr_sprint(ip, sizeof(ip), (struct sockaddr *) &ss);
            debug("child %lu for %s", (unsigned long) child, ip);
        }
    } while (1);
}


/*
 * Main routine.  Parses command-line arguments, determines whether we're
 * running in stand-alone or inetd mode, and does the connection handling if
 * running in standalone mode.  User connections are handed off to
 * process_connection.
 */
int
main(int argc, char *argv[])
{
    struct options options;
    FILE *pid_file;
    int option;
    struct sigaction sa;
    gss_cred_id_t creds = GSS_C_NO_CREDENTIAL;
    OM_uint32 minor;
    struct config *config;

    /*
     * Since we are normally called from tcpserver or inetd, prevent clients
     * from holding on to us forever by dying after an hour.
     */
    alarm(60 * 60);

    /* Ignore SIGPIPE errors from our children. */
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = SIG_IGN;
    if (sigaction(SIGPIPE, &sa, NULL) < 0)
        sysdie("cannot set SIGPIPE handler");

    /* Establish identity. */
    message_program_name = "remctld";

    /* Initialize options. */
    memset(&options, 0, sizeof(options));
    options.port = 4373;
    options.service = NULL;
    options.pid_path = NULL;
    options.config_path = CONFIG_FILE;

    /* Parse options. */
    while ((option = getopt(argc, argv, "dFf:hk:mP:p:Ss:v")) != EOF) {
        switch (option) {
        case 'd':
            options.debug = true;
            break;
        case 'F':
            options.foreground = true;
            break;
        case 'f':
            options.config_path = optarg;
            break;
        case 'h':
            usage(0);
            break;
        case 'k':
            if (setenv("KRB5_KTNAME", optarg, 1) < 0)
                sysdie("cannot set KRB5_KTNAME");
            break;
        case 'm':
            options.standalone = true;
            break;
        case 'P':
            options.pid_path = optarg;
            break;
        case 'p':
            options.port = atoi(optarg);
            break;
        case 'S':
            options.log_stdout = true;
            break;
        case 's':
            options.service = optarg;
            break;
        case 'v':
            printf("remctld %s\n", PACKAGE_VERSION);
            exit(0);
            break;
        default:
            usage(1);
            break;
        }
    }

    /* Daemonize if told to do so. */
    if (options.standalone && !options.foreground)
        daemon(0, options.log_stdout);

    /*
     * Set up syslog unless stdout/stderr was requested.  Set up debug logging
     * if requestsed.
     */
    if (options.log_stdout) {
        if (options.debug)
            message_handlers_debug(1, message_log_stdout);
    } else {
        openlog("remctld", LOG_PID | LOG_NDELAY, LOG_DAEMON);
        message_handlers_notice(1, message_log_syslog_info);
        message_handlers_warn(1, message_log_syslog_warning);
        message_handlers_die(1, message_log_syslog_err);
        if (options.debug)
            message_handlers_debug(1, message_log_syslog_debug);
    }

    /* Read the configuration file. */
    config = server_config_load(options.config_path);
    if (config == NULL)
        die("cannot read configuration file %s", options.config_path);

    /*
     * If a service was specified, we should load only those credentials since
     * those are the only ones we're allowed to use.  Otherwise, creds will
     * keep its default value of GSS_C_NO_CREDENTIAL, which means support
     * anything that's in the keytab.
     */
    if (options.service != NULL) {
        if (!acquire_creds(options.service, &creds))
            die("unable to acquire creds, aborting");
    }

    /*
     * Set up our PID file now after we've daemonized, since we may have
     * changed PIDs in the process.
     */
    if (options.standalone && options.pid_path != NULL) {
        pid_file = fopen(options.pid_path, "w");
        if (pid_file == NULL)
            sysdie("cannot create PID file %s", options.pid_path);
        fprintf(pid_file, "%ld\n", (long) getpid());
        fclose(pid_file);
    }

    /*
     * If we're not running as a daemon, just process the connection.
     * Otherwise, create a socket and listen on the socket, processing each
     * incoming connection.
     */
    if (!options.standalone)
        server_handle_connection(0, config, creds);
    else
        server_daemon(&options, config, creds);

    /* Clean up and exit.  We only reach here in regular mode. */
    if (creds != GSS_C_NO_CREDENTIAL)
        gss_release_cred(&minor, &creds);
    return 0;
}
